### PR TITLE
Remove unused/wrong loaded vcls

### PR DESCRIPTION
**Pull request by awmackowiak**

### CODE REVIEW

File: vaas-app/src/vaas/cluster/cluster.py [hunk 1]
```diff
@@ -59,7 +59,7 @@ def load_vcl(self, start_processing_time, clusters):
                 cluster.reload_timestamp = start_processing_time
                 cluster.save()
         except VclLoadException as e:
-            self.logger.error('Loading error: {} - rendered vcl-s not used'.format(e))
+            parallel_loader.discard_loaded_unused_vcl(vcl_list)
```
Comment (ziollek) [inline vaas-app/src/vaas/cluster/cluster.py]: In this way we loose information about connectivity problems with varnish (look at line #78)

File: vaas-app/src/vaas/cluster/cluster.py [hunk 2]
```diff
@@ -154,6 +154,10 @@ def _append_vcl(self, vcl, server, future_results, executor):
         loader = VclLoader(self.api_provider.get_api(server), server.status == 'maintenance')
         future_results.append(tuple([vcl, loader, server, executor.submit(loader.load_new_vcl, vcl)]))
 
+    def _discard_vcl(self, server, future_results, executor):
+        loader = VclLoader(self.api_provider.get_api(server), server.status == 'maintenance')
```
Comment (ziollek) [inline vaas-app/src/vaas/cluster/cluster.py]: Try to eliminate code duplication (creating instance of VclLoader - could be done once)

File: vaas-app/src/vaas/cluster/cluster.py [hunk 3]
```diff
@@ -204,6 +211,23 @@ def use_vcl_list(self, vcl_name, vcl_loaded_list):
 
         return result
 
+    def discard_loaded_unused_vcl(self, vcl_list):
+        start = time.time()
+        future_results = []
+        result = True
+        with ThreadPoolExecutor(max_workers=self.max_workers) as executor:
+            for server, _ in vcl_list:
+                self._discard_vcl(server, future_results, executor)
```
Comment (ziollek) [inline vaas-app/src/vaas/cluster/cluster.py]: Pls explicite change future_results ie.:
future_results.append(self._discard_vcl(server, executor))

File: vaas-app/src/vaas/cluster/tests/test_cluster.py [hunk 4]
```diff
@@ -239,11 +239,12 @@ def test_should_not_use_vcls_on_error_while_loading_vcl(self):
...
             with patch.object(ParallelLoader, 'load_vcl_list', side_effect=VclLoadException):
-                with patch.object(ParallelLoader, 'use_vcl_list', return_value=False) as use_vcl_mock:
-                    varnish_cluster = VarnishCluster()
-                    with self.assertRaises(VclLoadException):
-                        varnish_cluster.load_vcl(timezone.now(), [])
-                    """
-                    Here we check if 'use' command is NOT sent to servers
-                    """
-                    assert_list_equal([], use_vcl_mock.call_args_list)
+                with patch.object(ParallelLoader, 'discard_loaded_unused_vcl', side_effect=VclLoadException):
```
Comment (ziollek) [inline vaas-app/src/vaas/cluster/tests/test_cluster.py]: one test = one behaviour. pls add another test: test_should_discard_unused_vcls_on_error_while_loading_vcl

File: vaas-app/src/vaas/cluster/cluster.py [hunk 5]
```diff
@@ -60,6 +57,7 @@ def load_vcl(self, start_processing_time, clusters):
                 cluster.save()
         except VclLoadException as e:
             self.logger.error('Loading error: {} - rendered vcl-s not used'.format(e))
+            parallel_loader.discard_loaded_unused_vcl(vcl_list)
```
Comment (ziollek) [inline vaas-app/src/vaas/cluster/cluster.py]: what happens if error occure while discarding unused vcl ?

File: vaas-app/src/vaas/cluster/cluster.py [hunk 6]
```diff
@@ -148,12 +146,17 @@ class ParallelLoader(ParallelExecutor):
     def __init__(self, max_workers=10):
         ParallelExecutor.__init__(self, max_workers)
         self.api_provider = VarnishApiProvider()
+        self.loader_handler = VclLoader
 
     def _append_vcl(self, vcl, server, future_results, executor):
         """Suppress exceptions if cannot load vcl for server in maintenance state"""
-        loader = VclLoader(self.api_provider.get_api(server), server.status == 'maintenance')
+        loader = self.loader_handler(self.api_provider.get_api(server), server.status == 'maintenance')
```
Comment (ziollek) [inline vaas-app/src/vaas/cluster/cluster.py]: You just assing class name to variable (#149) - creating objects is still duplicated

File: vaas-app/src/vaas/cluster/tests/test_cluster.py [hunk 7]
```diff
@@ -239,11 +239,25 @@ def test_should_not_use_vcls_on_error_while_loading_vcl(self):
...
-                with patch.object(ParallelLoader, 'use_vcl_list', return_value=False) as use_vcl_mock:
+                with patch.object(ParallelLoader, 'discard_loaded_unused_vcl', side_effect=VclLoadException):
+                    with patch.object(ParallelLoader, 'use_vcl_list', return_value=False) as use_vcl_mock:
+                        varnish_cluster = VarnishCluster()
+                        with self.assertRaises(VclLoadException):
+                            varnish_cluster.load_vcl(timezone.now(), [])
+                        """
+                        Here we check if 'use' command is NOT sent to servers
+                        """
+                        assert_list_equal([], use_vcl_mock.call_args_list)
```
Comment (ziollek) [inline vaas-app/src/vaas/cluster/tests/test_cluster.py]: do not duplicate previous assertions

File: vaas-app/src/vaas/cluster/tests/test_cluster.py [hunk 8]
```diff
@@ -239,11 +239,25 @@ def test_should_not_use_vcls_on_error_while_loading_vcl(self):
 
         with patch.object(ParallelRenderer, 'render_vcl_for_servers', return_value=rendered_list):
             with patch.object(ParallelLoader, 'load_vcl_list', side_effect=VclLoadException):
-                with patch.object(ParallelLoader, 'use_vcl_list', return_value=False) as use_vcl_mock:
+                with patch.object(ParallelLoader, 'discard_loaded_unused_vcl', side_effect=VclLoadException):
```
Comment (ziollek) [inline vaas-app/src/vaas/cluster/tests/test_cluster.py]: really? exception should be raised also by discard ?

File: vaas-app/src/vaas/cluster/tests/test_cluster.py [hunk 9]
```diff
@@ -191,7 +191,7 @@ def test_should_discard_old_vcls(self):
         vcl_loaded_list = [(first_vcl, loader_mock, servers[1])]
 
         ParallelLoader().use_vcl_list('test', vcl_loaded_list)
-
+        print(loader_mock.discard_unused_vcls.call_args_list)
```
Comment (ziollek) [inline vaas-app/src/vaas/cluster/tests/test_cluster.py]: do not print anything in tests - keep this secret only for you ;)

File: vaas-app/src/vaas/vcl/loader.py [hunk 10]
```diff
@@ -71,9 +72,9 @@ def discard_unused_vcls(self):
                 else:
                     return_value = max(return_value, VclStatus.ERROR.value)
                     self.logger.warning("VCL %s not discarded." % vcl)
-            except AssertionError:
-                return_value = max(return_value, VclStatus.ERROR.value)
+            except AssertionError as e:
                 self.logger.warning("VCL %s not discarded." % vcl)
+                raise VclDiscardException(e)
```
Comment (ziollek) [inline vaas-app/src/vaas/vcl/loader.py]: You change original behaviour - previously we can try to discard all vcl's independent of any errors - now you break everything on first problem - this is your intention ?

File: vaas-app/src/vaas/cluster/tests/test_cluster.py (hunk 9 above)
Comment (ziollek) [inline vaas-app/src/vaas/cluster/tests/test_cluster.py]: remove print

**Comment by jtaczanowski** [comment]
+1

**Pull request merged**